Throw an error in `bitTest*` if the bit position is out-of-bounds

pamarcos opened the pull request:
Fixes #65517

Throw an error for bit positions out of bounds for the `bitTest` function family. This is also [what other SQL servers do](https://learn.microsoft.com/en-us/sql/t-sql/functions/get-bit-transact-sql?view=sql-server-ver16).

In the case of `bitTestAny` the logic behind wouldn't be affected if a user includes bit positions out of bounds. But for consistency, I think we'd rather have the same behavior everywhere.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `bitTest`, `bitTestAll`, and `bitTestAny` now return an error if the specified bit index is out-of-bounds (issue #65517)

pamarcos committed:
Throw exception in bitTest when position is out of bound

This happens whenever the number of bit positions is bigger
than the number of bits in the number, or when the bit position
is negative.

pamarcos committed:
Add bound check to bitTestAll and bitTestAny

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #65818 from pamarcos/ClickHouse:bittest-throw-error-if-out-of-bounds -->
*This is an automated comment for commit d63f70e4f1459bbb7808fd1fb7202e9648ad6570 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65818/d63f70e4f1459bbb7808fd1fb7202e9648ad6570/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/f3997640211ba3b29213191b9d6e194536233399/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos commented:
I removed the ubsan test for bittest since I'm making the behavior well-defined, but pinging @alexey-milovidov since he's the one [who introduced it](https://github.com/ClickHouse/ClickHouse/commit/f748efbb5b234d8dfe9c587edf8d79d606f2aba5) in case there's something I'm missing 🙏

alexey-milovidov commented:
Looks good. I hope it will not introduce performance regressions.

PS. We allow implementation-defined behavior in many other functions when it depends on the CPU or on compilation flags in out-of-bound cases but do not introduce anything incorrect in other places. When the compiler knows at compile time that there is undefined behavior, it can think that it is impossible and apply incorrect optimizations in other places. But if it is a runtime parameter, it is ok to mark the function as NO_SANITIZE_UNDEFINED.

rschu1ze commented on src/Functions/FunctionBitTestMany.h:
```diff
@@ -17,6 +17,7 @@ namespace ErrorCodes
     extern const int ILLEGAL_COLUMN;
     extern const int ILLEGAL_TYPE_OF_ARGUMENT;
     extern const int TOO_FEW_ARGUMENTS_FOR_FUNCTION;
+    extern const int PARAMETER_OUT_OF_BOUND;
```
Nit: Let's keep sorted lists sorted.

rschu1ze commented on src/Functions/bitTest.cpp:
```diff
@@ -21,12 +22,21 @@ struct BitTestImpl
         else
-            return (typename NumberTraits::ToInteger<A>::Type(a) >> typename NumberTraits::ToInteger<B>::Type(b)) & 1;
+        {
+            typename NumberTraits::ToInteger<A>::Type a_int(a);
```
Nit: assignment initialization is more aesthetic:

```cpp
typename NumberTraits::ToInteger<A>::Type a_int = a;
typename NumberTraits::ToInteger<B>::Type b_int = b;
```

Also, `static_cast` is a little bit more C++-ish (l. 33): `static_cast<decltype(b)>((8 * sizeof(a)) - 1);`

rschu1ze commented on src/Functions/bitTest.cpp:
```diff
@@ -21,12 +22,21 @@ struct BitTestImpl
+            typename NumberTraits::ToInteger<A>::Type a_int(a);
+            typename NumberTraits::ToInteger<B>::Type b_int(b);
+            const auto max_position = decltype(b)((8 * sizeof(a)) - 1);
+            if (b > max_position || b < 0)
```
Nit: Suggest to do the check (l. 33-36) with `a_int` / `b_int` (instead of `a`/`b`) or keep it as is and move it up before l. 31.

rschu1ze commented on tests/queries/0_stateless/01082_bit_test_out_of_bound.sql:
```diff
@@ -1,2 +1,9 @@
+SELECT number, bitTestAll(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8);
+SELECT number, bitTestAll(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8, 16); -- { serverError PARAMETER_OUT_OF_BOUND }
+
+SELECT number, bitTest(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8);
```
General tip about tests: In case of problems with the test, it can be difficult to find out how the data in the expected results file relates to the queries in the SQL test. Here, the former is a bunch of numbers...

"Separator SELECTs" can help with that, e.g.

```sql
SELECT '-- bitTestAny';
SELECT number, bitTestAny(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8);
SELECT number, bitTestAny(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8, 16); -- { serverError PARAMETER_OUT_OF_BOUND }

SELECT '-- bitTestAll';
SELECT number, bitTestAny(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8);
SELECT number, bitTestAny(toUInt8(1 + 4 + 16 + 64), number) FROM numbers(8, 16); -- { serverError PARAMETER_OUT_OF_BOUND }

[...]
```

rschu1ze commented on tests/queries/0_stateless/01710_minmax_count_projection.sql:
```diff
@@ -16,7 +16,7 @@ select min(i), max(i), count() from d where _partition_value.1 = 10 group by _pa
 
 -- fuzz crash https://github.com/ClickHouse/ClickHouse/issues/37151
-SELECT min(i), max(i), count() FROM d WHERE (_partition_value.1) = 0 GROUP BY ignore(bitTest(ignore(NULL), 65535), NULL, (_partition_value.1) = 7, '10.25', bitTest(NULL, -9223372036854775808), NULL, ignore(ignore(-2147483647, NULL)), 1024), _partition_id ORDER BY _partition_id ASC NULLS FIRST;
+SELECT min(i), max(i), count() FROM d WHERE (_partition_value.1) = 0 GROUP BY ignore(bitTest(ignore(NULL), 0), NULL, (_partition_value.1) = 7, '10.25', bitTest(NULL, 0), NULL, ignore(ignore(-2147483647, NULL)), 1024), _partition_id ORDER BY _partition_id ASC NULLS FIRST;
```
While at it, you probably want to fix the docs, it says `The countdown starts from 0 from the right to the left.`, it should say: `Counting is right-to-left, starting at 0.`

pamarcos committed:
Address issues pointed out in the PR

pamarcos commented:
@rschu1ze thanks a lot for the detailed review :heart: 
I addressed all issues in https://github.com/ClickHouse/ClickHouse/pull/65818/commits/93c1b5d8a72559517bafd45cf94e2e531e39a404

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into bittest-throw-error-if-out-of-bounds

rschu1ze commented:
ClickHouse Upgrade Check (debug): Settings changes history not updated, unrelated.
ClickHouse Stateless Tests (debug, S3 Storage) [6/6]: #65877
